### PR TITLE
Added possibility to use useTranlation returning Array or Object

### PR DESCRIPTION
**Pull request by adrai**
Added possibility to use useTranlation not only expecting an array but also an object. Should help #712

### CODE REVIEW

File: src/index.d.ts [hunk 1]
```diff
@@ -42,10 +42,16 @@ export function useSSR(initialI18nStore: any, initialLanguage: any): void;
 export interface UseTranslationOptions {
   i18n?: i18next.i18n;
 }
+export interface UseTranslationResponse extends Array<i18next.TFunction | i18next.i18n> {
```
Comment (adrai) [inline src/index.d.ts]: @rosskevin @schettino I've no clue about typescript... Is this ok?

**Comment by jamuhl** [comment]
@kachkaev opinions about letting users choose? we could provide both options

**Comment by kachkaev** [comment]
What if there is `const i18n = useI18n()?` as an alternative? It can be used as const { t } = useI18n()` if `t` is bound to `i18n` correctly (no issues with `this`)? The original hook can be marked as deprecated over time if it’s perceived as a less convenient one.

**Comment by jamuhl** [comment]
@kachkaev i want to avoid another hook. As 95% of the times you like to get some t function based on a given namespace i think this is prefered

```js
const { t } = useTranslation('ns1');
const [t] = useTranslation('ns1');
```

The only case you like to get the i18n instance is where you need current language or like to access the changeLanguage function. Chances are high you need t function in this component too.

So your approach would be:

```js
const i18n = useI18n();
const { t } = i18n;
```

while you could just:

```js
const { t, i18n } = useTranslation();
```

Beside that what we do for HOC and render prop? withI18n?!?

```js
function Extended({ i18n }) {
   const { t } = i18n; 
}
```

File: src/index.d.ts (hunk 1 above)
Comment (kachkaev) [reply src/index.d.ts]: @adrai you can try:

```ts
type UseTranslationResponse = [i18next.TFunction, i18next.i18n] & {
  t: i18next.TFunction;
  i18n: i18next.i18n;
}
```

**Comment by kachkaev** [comment]
Oh, I did not realise that JavaScript would allow us to return something that's both an array and an object! In this case, keeping just `useTranslation` makes perfect sense!

File: src/index.d.ts (hunk 1 above)
Comment (adrai) [reply src/index.d.ts]: thx, just updated it

**Comment by adrai** [comment]
Yes, JavaScript allows to extend arrays.
With this PR you can now choose to do:
`const [ t, i18n ] = useTranslation();`
or
`const { t, i18n } = useTranslation();`

**Comment by kachkaev** [comment]
@jamuhl WDYT of updating docs too and encourage users to stick with `{ t, i18n }`? The array can still be still supported for BC, but this will be a 'hidden feature'.

File: test/typescript/examples.test.tsx [hunk 2]
```diff
@@ -22,6 +22,7 @@ function MyComponent() {
 // page uses the hook
 function Page() {
   const [t, i18n] = useTranslation();
+  // const { t, i18n } = useTranslation();
```
Comment (rosskevin) [inline test/typescript/examples.test.tsx]: please remove comment

Comment (rosskevin) [reply test/typescript/examples.test.tsx]: Actually, we need both if you want to merge this PR.  If you are expecting it to return both.

Comment (rosskevin) [reply test/typescript/examples.test.tsx]: I'll push some test changes.

**Comment by rosskevin** [comment]
> @jamuhl WDYT of updating docs too and encourage users to stick with `{ t, i18n }`? The array can still be still supported for BC, but this will be a 'hidden feature'.

I have added explicit TS tests for both usages, and I agree that the object usage should be promoted as the default (my commit reflects that) especially for js users not using types.

File: src/index.d.ts (hunk 1 above)
Comment (rosskevin) [reply src/index.d.ts]: This is bizarre - but I confirmed it typechecks correctly.

**Comment by jamuhl** [comment]
@rosskevin @kachkaev @adrai yes...i will update the docs / samples in favor to using object spread `const { t, i18n } = useTranslation();`

**Comment by jamuhl** [comment]
published in react-i18next@10.0.1

**Comment by jamuhl** [comment]
https://react.i18next.com/latest/usetranslation-hook and all the other places in the docs should reflect this change

**Pull request merged**